core/retries/src/main/java/software/amazon/awssdk/retries/internal/ratelimiter/RateLimiterTokenBucket.java

```diff
@@ -0,0 +1,287 @@
+ * <a href="https://en.wikipedia.org/wiki/CUBIC_TCP">CUBIC</a>.
+ */
+@SdkInternalApi
+public class RateLimiterTokenBucket {
+    private final AtomicReference<PersistentState> stateReference;
+    private final RateLimiterClock clock;
+
+    RateLimiterTokenBucket(RateLimiterClock clock) {
+        this.clock = clock;
+        this.stateReference = new AtomicReference<>(new PersistentState());
+    }
+
+    /**
+     * Acquires one token from the rate limiter and returns a response with a delay value (possibly zero). Callers are expected to
+     * wait for at least this amount of time before retrying the request. See {@link RateLimiterAcquireResponse#delay()}.
+     */
+    public RateLimiterAcquireResponse tryAcquire() {
+        StateUpdate<Duration> update = updateState(ts -> ts.tokenBucketAcquire(clock, 1.0));
+        return RateLimiterAcquireResponse.create(update.result);
+    }
+
+    /**
+     * Updates the estimated send rate after a throttling response.
+     */
+    public RateLimiterUpdateResponse updateRateAfterThrottling() {
+        StateUpdate<Void> update = consumeState(ts -> ts.updateClientSendingRate(clock, true));
+        return RateLimiterUpdateResponse.builder()
+            .measuredTxRate(update.newState.measuredTxRate())
+            .fillRate(update.newState.fillRate())
+            .build();
+    }
+
+    /**
+     * Updates the estimated send rate after a successful response.
+     */
+    public RateLimiterUpdateResponse updateRateAfterSuccess() {
+        StateUpdate<Void> update = consumeState(ts -> ts.updateClientSendingRate(clock, false));
+        return RateLimiterUpdateResponse.builder()
+                                        .measuredTxRate(update.newState.measuredTxRate())
+                                        .fillRate(update.newState.fillRate())
+                                        .build();
+    }
+
+    private StateUpdate<Void> consumeState(Consumer<TransientState> mutator) {
+        return updateState(ts -> {
+            mutator.accept(ts);
+            return null;
+        });
+    }
+
+    private <T> StateUpdate<T> updateState(Function<TransientState, T> mutator) {
+        PersistentState current;
+        PersistentState updated;
+        T result;
+        do {
+            current = stateReference.get();
+            TransientState transientState = current.toTransient();
+            result = mutator.apply(transientState);
+            updated = transientState.toPersistent();
+        } while (!stateReference.compareAndSet(current, updated));
+
+        return new StateUpdate<>(updated, result);
+    }
+
+    static class StateUpdate<T> {
+        private final PersistentState newState;
+        private final T result;
+
+        StateUpdate(PersistentState newState, T result) {
+            this.newState = newState;
+            this.result = result;
+        }
+    }
+
+    private static class TransientState {
```

Comment:
It would be really helpful for readability if we could have APIs related to tokenBucket in RateLimiterTokenBucket class and other methods in a Helper class

Reply:
Not sure what you mean by "other methods", all the code contained in this class is related to the rate limiter token bucket. Can you please point me which methods you think should go into a different class?

Reply:
I was trying to say if we have RateLimiterTokenBucket with basic its APIs to acquireTokens and refillToken and rest of the other functionalities like refill based on Success, exception or throtlle exception should be done by another wrapping or helper class.
That we can limit RateLimiterTokenBucket responsibility to just acquire and refill and the helper class will acquire/refill based on the success failure scenarios

Reply:
Other methods are methods that updates the tokens based on error/success criteria like updateRateAfterThrottling etc